src/main/java/net/neoforged/neoforge/registries/NewRegistryEvent.java

```diff
@@ -5,108 +5,74 @@
-     * @param onFill  Called when the returned supplier is filled with the registry
-     * @return a supplier of the {@link IForgeRegistry} created by the builder. Resolving too early will return null.
+     * @param registry the registry to register
      */
-    public <V> Supplier<IForgeRegistry<V>> create(RegistryBuilder<V> builder, @Nullable Consumer<IForgeRegistry<V>> onFill) {
-        RegistryHolder<V> registryHolder = new RegistryHolder<>();
-
-        registries.add(new RegistryData<>(builder, registryHolder, onFill));
-
-        return registryHolder;
+    public <T> void register(Registry<T> registry) {
+        this.registries.add(registry);
     }
 
-    @SuppressWarnings("deprecation")
     void fill() {
-        RuntimeException aggregate = new RuntimeException();
-        Map<RegistryBuilder<?>, IForgeRegistry<?>> builtRegistries = new IdentityHashMap<>();
-
-        if (BuiltInRegistries.REGISTRY instanceof MappedRegistry<?> rootRegistry)
-            rootRegistry.unfreeze();
+        ((BaseMappedRegistry<?>) BuiltInRegistries.REGISTRY).unfreeze();
 
-        for (RegistryData<?> data : this.registries) {
-            try {
-                buildRegistry(builtRegistries, data);
-            } catch (Throwable t) {
-                aggregate.addSuppressed(t);
-                return;
-            }
+        for (final var registry : this.registries) {
+            registerToRootRegistry(registry);
         }
 
-        if (BuiltInRegistries.REGISTRY instanceof MappedRegistry<?> rootRegistry)
-            rootRegistry.freeze();
-
-        if (aggregate.getSuppressed().length > 0)
-            LOGGER.error(LogUtils.FATAL_MARKER, "Failed to create some forge registries, see suppressed exceptions for details", aggregate);
-    }
-
-    private <T> void buildRegistry(Map<RegistryBuilder<?>, IForgeRegistry<?>> builtRegistries, RegistryData<T> data) {
-        RegistryBuilder<T> builder = data.builder;
-        IForgeRegistry<T> registry = builder.create();
-
-        builtRegistries.put(builder, registry);
-
-        if (builder.getHasWrapper() && !BuiltInRegistries.REGISTRY.containsKey(registry.getRegistryName()))
-            RegistryManager.registerToRootRegistry((ForgeRegistry<?>) registry);
-
-        data.registryHolder.registry = registry;
-        if (data.onFill != null)
-            data.onFill.accept(registry);
+        ((WritableRegistry<?>) BuiltInRegistries.REGISTRY).freeze();
     }
 
-    private record RegistryData<V>(
-            RegistryBuilder<V> builder,
-            RegistryHolder<V> registryHolder,
-            Consumer<IForgeRegistry<V>> onFill) {}
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    private void registerToRootRegistry(Registry<?> registry) {
+        ResourceLocation registryName = registry.key().location();
+        if (BuiltInRegistries.REGISTRY.containsKey(registryName))
+            throw new IllegalStateException("Tried to register new registry for " + registryName);
```

Comment:
The exception message doesn't seem... descriptive of the error. Perhaps "Attempted duplicate registration of registry " + registryName?

Reply:
Fixed in f2801cb37.